autotools: Clean up after adding Wycheproof

real-or-random opened the pull request:
Follow-up to https://github.com/bitcoin-core/secp256k1/pull/1245.

This builds on top of https://github.com/bitcoin-core/secp256k1/pull/1276. Let's only merge https://github.com/bitcoin-core/secp256k1/pull/1276 as a hotfix for the Core build.

hebasto commented on Makefile.am:
```diff
@@ -217,11 +218,29 @@ precomp: $(PRECOMP)
+	mkdir -p $(@D)
+	python3 tools/tests_wycheproof_generate.py src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.json > $@
+
+testvectors: $(TESTVECTORS)
```
```suggestion
.PHONY: testvectors
testvectors: $(TESTVECTORS)
```

hebasto replied:
nm, ignore this ^

hebasto reviewed (commented):
In 3955f8f9940f35a37faa01c08d17795adf8ceb43
> autotools: Rename "pregenerated" clean targets and make them .PHONY
> 
> This follows the automake conventions, see:
> https://www.gnu.org/software/automake/manual/html_node/Clean.html

The `clean-precomp` was `.PHONY`: https://github.com/bitcoin-core/secp256k1/blob/3bab71cf0534fd5948386de2b6ac31a7044ad76b/Makefile.am#L1

A comment needs to be updated: https://github.com/bitcoin-core/secp256k1/blob/3bab71cf0534fd5948386de2b6ac31a7044ad76b/Makefile.am#L205

Why not just `MAINTAINERCLEANFILES = $(PRECOMP) $(TESTVECTORS)`?

hebasto commented:
```diff
--- a/ci/cirrus.sh
+++ b/ci/cirrus.sh
@@ -109,7 +109,7 @@ fi
 # Rebuild precomputed files (if not cross-compiling).
 if [ -z "$HOST" ]
 then
-    make clean-precomp clean-testvectors
+    make maintainer-clean
     make precomp testvectors
 fi
 
```

to fix CI?

real-or-random commented:
> Why not just `MAINTAINERCLEANFILES = $(PRECOMP) $(TESTVECTORS)`?

Yeah, that would work, too. But I think it's nice to have the ability to clean `precomp` or `testvectors` separately. I'll address the other comments.

(Apparently it was a good idea that I opened a separate PR for the first commit.)

real-or-random commented:
I changed my mind and removed the renaming. It's not clear that a "maintainer"-prefix is better... Sorry for the back and forth.

Independently, `make maintainer-clean` in the Cirrus script won't work because that target also removes the Makefile.

hebasto commented on Makefile.am:
```diff
@@ -217,11 +216,29 @@ precomp: $(PRECOMP)
+TESTVECTORS = src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.h
+
+src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.h:
+	mkdir -p $(@D)
```
For more robustness, could consider
```suggestion
	$(MKDIR_P) $(@D)
```

with the [`AC_PROG_MKDIR_P`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/autoconf.html#index-AC_005fPROG_005fMKDIR_005fP-1) macro in the `configure.ac`?

real-or-random replied:
As I understand the history of this, the reason why this macro exists is that some implementations of `mkdir -p` on some old systems are not thread safe. And that's a problem in a parallel make, where multiple make processes try to create the same directory (see https://lists.gnu.org/r/automake/2006-04/msg00084.html).

I don't see how this could be a problem here, given that there's only a single recipe in the Makefile that creates this directory. That's why I think it's better to stick to `mkdir -p` here for simplicity.

hebasto commented on Makefile.am:
```diff
@@ -217,11 +216,29 @@ precomp: $(PRECOMP)
+
+src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.h:
+	mkdir -p $(@D)
+	python3 tools/tests_wycheproof_generate.py src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.json > $@
```
2dd672a287a0c9c4382ffbf56f690727f2e74696

To make it work in a VPATH build, the diff required as follows:
```suggestion
	python3 $(top_srcdir)/tools/tests_wycheproof_generate.py $(top_srcdir)/src/wycheproof/ecdsa_secp256k1_sha256_bitcoin_test.json > $@
```

real-or-random replied:
added a commit

hebasto commented on Makefile.am:
```diff
@@ -202,7 +201,7 @@ precompute_ecmult_gen_LDADD = $(COMMON_LIB)
 # build by a normal user) depends on mtimes, and thus is very fragile.
 # This means that rebuilds of the prebuilt files always need to be
-# forced by deleting them, e.g., by invoking `make clean-precomp`.
+# forced by deleting them, e.g., using `make maintainer-clean`.
```
Considering that `make maintainer-clean` [removes](https://github.com/bitcoin-core/secp256k1/pull/1277#issuecomment-1514815590) `Makefile`, could this comment be confusing if someone follows it blindly?

real-or-random replied:
I just removed that part.

hebasto reviewed (approved):
ACK 7e977b3c5071fc17575ff88ebbc9db7b17c70497
